frontend/model/state.js

```diff
@@ -22,6 +23,39 @@ import identityGetters from './contracts/shared/getters/identity.js'
 import notificationModule from '~/frontend/model/notifications/vuexModule.js'
 import settingsModule from '~/frontend/model/settings/vuexModule.js'
 import chatroomModule from '~/frontend/model/chatroom/vuexModule.js'
+import { CONTRACTS_MODIFIED } from '../../shared/domains/chelonia/events.js'
+
+// Wrapper function for performing contract upgrades and migrations
+const contractUpdate = (updateFn: (contractIDHints: ?string[]) => any) => {
+  const loginErrorHandler = () => {
+    sbp('okTurtles.events/off', CONTRACTS_MODIFIED, modifiedHandled)
+    sbp('okTurtles.events/off', LOGIN_COMPLETE, loginCompleteHandler)
+    sbp('okTurtles.events/off', LOGOUT, logoutHandler)
+  }
+  const logoutHandler = () => {
+    sbp('okTurtles.events/off', CONTRACTS_MODIFIED, modifiedHandled)
+  }
+  const loginCompleteHandler = () => {
+    sbp('okTurtles.events/off', LOGIN_ERROR, loginErrorHandler)
+  }
+  // This function is called when the set of subscribed contracts is modified
+  const modifiedHandled = (_, { added }) => {
+    // Wait for the added contracts to be ready, then call the update function
+    sbp('chelonia/contract/wait', added).then(() => {
+      updateFn(added)
+    })
+  }
+
+  // Add event listeners for CONTRACTS_MODIFIED, LOGOUT, LOGIN_COMPLETE and
+  // LOGIN_ERROR
+  sbp('okTurtles.events/on', CONTRACTS_MODIFIED, modifiedHandled)
+  sbp('okTurtles.events/once', LOGOUT, logoutHandler)
+  sbp('okTurtles.events/once', LOGIN_COMPLETE, loginCompleteHandler)
+  sbp('okTurtles.events/once', LOGIN_ERROR, loginErrorHandler)
+
+  // Call the update function in the next tick
+  setTimeout(updateFn, 0)
+}
```

Comment:
> Is there someway to make it so that after the upgrade gets run, it's the modifiedHandled is no longer called?

One idea that occurred to me, is that `updateFn` can pass in a function that will unregister the listener that can be called once a successful upgrade is done.

Reply:
> Is there someway to make it so that after the upgrade gets run, it's the modifiedHandled is no longer called?

Well, no, because that's the entire point of this change not to have to refresh the page.

Reply:
But some of these functions only need to be run once, right? No need to call them every single time a contract is added or removed.

Did you see this?

> One idea that occurred to me, is that `updateFn` can pass in a function that will unregister the listener that can be called once a successful upgrade is done.

Reply:
> This seems like a potentially useful utility function. Can we import it from some other location instead of here? (Is there a way to make this available as a Chelonia dev utility)?

Useful for what, other than upgrades?

> Is there a way to make this available as a Chelonia dev utility

One issue with that is that all of the events other than `CONTRACTS_MODIFIED` are GI events, not Chelonia events.

Reply:
> Useful for what, other than upgrades? Like, I'm all for moving it, but I don't see it as being particularly useful as a generic function outside it being used here.

Useful for anyone writing Chelonia-based apps. This is one of 2 ways we've got for doing upgrades (and the only way that exists currently). Seems like every dev would need it.

> One issue with that is that all of the events other than CONTRACTS_MODIFIED are GI events, not Chelonia events.

Honestly, with how complicated logging in and logging out is, I think we will be forced to move that code into Chelonia.

Reply:
> with how complicated logging in and logging out is, I think we will be forced to move that code into Chelonia.

I'm not sure that feasible nor a good idea, and it's not really that logging in or out is complicated: it is, in a way, but that's because there are things that need to be done in specific places (the SW and the app), and because of specific things related to saving and loading the state, which can also need encryption and decryption.

In this case, the `LOGIN_COMPLETE` and `LOGIN_ERROR` events are also only on the application / browsing-context side (implementing this function in the actions / SW side wouldn't need these events, _maybe_ it'd need `LOGIN`, but you asked to use the existing selector we have).

Moving logging in and out into Chelonia means that certain types of contracts need to be defined, rather than having a generic system.

In another type of application, you only really need to listen for the `CONTRACTS_MODIFIED` modified event.

Reply:
Can we leave this without major changes and improve it later? Like I said, I'm happy to put it in a different file (in fact, I think it doesn't belong in this already complicated file). My question is more about the 'moving it to Chelonia' part, as this is feeling like an ever-growing PR.